src/vm/analysis/update_expressions_sorting/mod.rs

```diff
@@ -0,0 +1,216 @@
+        match expr.expr {
+            AtomValue(_) => Ok(true),
+            Atom(ref name) => {
+                if let Some(dep) = self.top_level_expressions_map.get(&name.clone()) {
+                    if dep.atom_index != expr.id {
+                        self.graph.add_directed_edge(tle_index, dep.expr_index)?;
+                    }
+                }
+                Ok(true)
+            },
+            List(ref exprs) => {
+                for expr in exprs.into_iter() {
+                    self.register_dependencies(expr, tle_index)?;
+                }
+                Ok(true)
+            }
+        }
+    }
+
+    fn find_expression_definition<'b>(&mut self, exp: &'b SymbolicExpression) -> Option<(String, u64, &'b SymbolicExpression)> {
+        if let Some(expression) = exp.match_list() {
+            if let Some((function_name, function_args)) = expression.split_first() {
+                if let Some(definition_type) = function_name.match_atom() {
+                    match definition_type.as_str() {
+                        "define-map" | "define-data-var" | "define" | "define-public" | "define-read-only" => {
+                            if function_args.len() > 1 {
+                                let define_expr = match function_args[0].match_list() {
+                                    Some(list) => &list[0],
+                                    _ => &function_args[0]
+                                };
+                                if let Some(tle_name) = define_expr.match_atom() {
+                                    return Some((tle_name.clone(), define_expr.id, define_expr));
+                                }   
+                            }
+                        }
+                        _ => {}
+                    }
+                } 
+            } 
+        }
+        None
+    }
+}
+
+pub struct TopLevelExpressionIndex {
+    expr_index: usize,
+    atom_index: u64
+}
+
+struct Graph {
+    nodes: Vec<usize>,
+    adjacency_list: Vec<Vec<usize>>
+}
+
+impl Graph {
+    fn new() -> Self {
+        Self { 
+            nodes: Vec::new(), 
+            adjacency_list: Vec::new() 
+        }
+    }
+
+    fn add_node(&mut self, expr_index: usize) -> CheckResult<()> {
+        self.nodes.push(expr_index);
+        self.adjacency_list.push(vec![]);
+        Ok(())
+    }
+
+    fn add_directed_edge(&mut self, src_expr_index: usize, dst_expr_index: usize) -> CheckResult<()> {
+        let list = self.adjacency_list.get_mut(src_expr_index).unwrap();
+        list.push(dst_expr_index);
+        Ok(())
+    }
+    
+    fn get_node_descendants(&self, expr_index: usize) -> Vec<usize> {
+        self.adjacency_list[expr_index].clone()
+    }
+
+    fn has_node_descendants(&self, expr_index: usize) -> bool {
+        self.adjacency_list[expr_index].len() > 0
+    }
+}
+
+struct GraphWalker {
+    seen: HashSet<usize>,
+}
+
+impl GraphWalker  {
```

Comment:
Would be useful to add a comment stating that this implements a depth-first search producing a post-order sort.

Reply:
Addressed in https://github.com/blockstack/blockstack-core/pull/1079/commits/c2154d41f52217e635087369bd605832838d7a78